fetch-in-react-practice/client/components/app.jsx

```diff
@@ -0,0 +1,111 @@
+/* eslint-disable no-console */
+import React from 'react';
+import PageTitle from './page-title';
+import TodoList from './todo-list';
+import TodoForm from './todo-form';
+
+class App extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      todos: []
+    };
+    this.addTodo = this.addTodo.bind(this);
+    this.toggleCompleted = this.toggleCompleted.bind(this);
+  }
+
+  componentDidMount() {
+    this.getAllTodos();
+  }
+
+  getAllTodos() {
+    /**
+     * Use fetch to send a GET request to `/api/todos`.
+     * Then 😉, once the response JSON is received and parsed,
+     * update state with the received todos.
+     */
+    fetch('/api/todos')
+      .then(res => res.json())
+      .then(res => this.setState({ todos: res }))
+      .catch(err => console.error(err.message));
+  }
+
+  addTodo(newTodo) {
+    /**
+     * Use fetch to send a POST request to `/api/todos`.
+     * Then 😉, once the response JSON is received and parsed,
+     * add the created todo to the state array.
+     *
+     * TIP: Be sure to SERIALIZE the todo in the body with JSON.stringify()
+     * And specify the "Content-Type" header as "application/json"
+     */
+    fetch('/api/todos', {
+      method: 'POST',
+      headers: {
+        'Content-Type': 'application/json'
+      },
+      body: JSON.stringify(newTodo)
+    })
+      .then(res => res.json())
+      .then(res => this.setState(state => {
+        const currentTodo = [...this.state.todos];
+        currentTodo[currentTodo.length] = res;
+        return { todos: currentTodo };
+      }))
+      .catch(err => console.error(err.message));
+  }
+
+  toggleCompleted(todoId) {
+    /**
+     * Find the index of the matching todo in the state array.
+     * Find its "isCompleted" status.
+     * Make a new Object containing the opposite "isCompleted" status.
+     * Use fetch to send a PATCH request to `/api/todos/${todoId}`
+     * Then 😉, once the response JSON is received and parsed,
+     * replace the old todo in state.
+     *
+     * TIP: Be sure to SERIALIZE the updates in the body with JSON.stringify()
+     * And specify the "Content-Type" header as "application/json"
+     */
+    let isCompletedByTodoId = null;
+    let indexByTodoId = null;
+
+    this.state.todos.forEach(todo => {
+      if (todo.id === todoId) {
+        isCompletedByTodoId = !todo.isCompleted;
+        indexByTodoId = this.state.todos.indexOf(todo);
+      }
+    });
+
+    fetch(`/api/todos/${todoId}`, {
+      method: 'PATCH',
+      headers: {
+        'Content-Type': 'application/json'
+      },
+      body: JSON.stringify({ isCompleted: isCompletedByTodoId })
+    })
+      .then(res => res.json())
+      .then(res => this.setState(state => {
+        const currentTodo = [...this.state.todos];
+        currentTodo[indexByTodoId].isCompleted = isCompletedByTodoId;
```

Comment:
The objects inside are not copied if you only copy the array. Instead of updating the property of the object, completely replace the object

```js
currentTodo[indexByTodoId] = res
```

Reply:
Got it, issue fixed 
7480105